documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,687 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+`SupportedStyles` is a bitmask containing all of the cursor styles that are supported by this backend. This must be queried before setting `Style` - the currently active cursor style. An exception should be thrown if an attempt is made to set `Style` to an unsupported style or multiple styles (i.e. multiple bits set).
+
+`Image` uses `RawImage` as-is from Silk.NET.Core, and when set to a non-null value implicitly sets `Style` to custom. As such, you must query `SupportedStyles` before using this property as well. Setting `Image` to `null` will set `Style` back to a standard cursor style, defined by the implementation. It is therefore recommended you set `Style` explicitly when disabling a custom cursor. Note that setting `Style` to a non-`Custom` value will also implicitly set this property to `null`. Setting `Mode` **to** `Custom` explicitly is undefined behaviour, as `Image` won't be set at the time of setting `Mode`.
+
+`SupportedFlags` is a bitmask containing other supported options for the cursor which can be mixed and matched if supported. This must be queried before setting `Flags` - the currently active options. An exception should be thrown if an attempt is made to set an unsupported flag on `Flags`. Unlike the other properties, `Flags` can have multiple bits set.
+
+`Flags` replaces `IsConfined`, which was undefined behaviour on platforms where it was not supported in 2.X - now with `SupportedFlags` you can query its support before setting it.
+
+### Enums
+
+```cs
+[Flags]
+public enum CursorModes
+{
+    Normal,
+    Hidden = 1 << 0,
+    Disabled = 1 << 1,
+    Raw = 1 << 2
+}
+```
+```cs
+[Flags]
+public enum CursorStyles
+{
+    Default,
+    Arrow = 1 << 0,
+    IBeam = 1 << 1,
+    Crosshair = 1 << 2,
+    Hand = 1 << 3,
+    HResize = 1 << 4,
+    VResize = 1 << 5,
+    Custom = 1 << 6,
+}
+```
+```cs
+[Flags]
+public enum CursorFlags
+{
+    None,
+    Confined = 1 << 0
+}
+```
+
+# Keyboard Input
+
+Once again, the interface is very simple.
+
+```cs
+public interface IKeyboard : IInputDevice
+{
+    KeyboardState State { get; }
+    void BeginInput();
+    void EndInput();
+    void SetClipboardText(string? clipboard);
+}
+```
+
+`State` is the device state as defined earlier.
+
+`BeginInput` starts recording textual input, bringing up the on-screen keyboard on platforms where this is needed (i.e. mobile). Unlike 2.0, this is now required given that textual input is now on a full-string basis rather than a per-char basis, and we don't want to record textual input unless the user asks us to.
+
+For instance, instead of a `KeyChar` event being raised every time a character is pressed or repeated, a `TextInput` event is raised as the string is being built alleviating the need for developers to build the string themselves, given that the primary use of this event is for GUI input.
```

Comment:
So if my app asks you to type something into a textbox, the app wouldn't get the text until I'm done typing it? Usually the textbox fills up as you type

Reply:
No (will clarify) you get the string as it’s being built i.e. if you’re typing “Hello” you will get “H” and then “He” and then “Hel” etc etc.

Reply:
That's not memory efficient and unnecessarily complicates doing any filtering on the characters. I can't really vouch for changing `KeyChar` to this

Reply:
The problem is, there is no way with an immutable state to actually get things like key repeats. What if you capture state, and between the next time you capture state there have been 5 repeats of a character? You will only get `c` - you have no idea how many times that character has been repeated. Now you could say "oh we'd just store the number of repeats", but that doesn't work either because what if you have 5 repeats _and then_ a different key press - you have lost an entire 5 characters of the string. We have to buffer it somehow, and it only makes sense for _textual_ input to buffer it into a string. **INFORMATIVE TEXT:** The memory inefficiency is also one of the primary reasons why I specified we don't capture char pressed outside of `BeginInput`/`EndInput`, in addition to the write-once-run-everywhere implications we're shooting for in 3.0.

Reply:
I don't understand why immutable states complicates this, wouldn't `KeyChar` still get called once for every char, including repeats?

Reply:
No, the events only get called when we detect a state change (see line 106 of the latest diff describing how `Update` works).

Reply:
I'm still baffled by this whole redesign. You mention that with it you're trying to solve problems or issues reported by users, can you make a list of these? Maybe that'll help me better understand the reasoning

Reply:
The main issue from a user perspective that we're solving is how event bindings work. Today, users complain about having to bind events on each individual device whereas realistically there's no reason for them to do this.

The Working Group indicated that it would like to be able to capture state, today this is a second class citizen thus we have redesigned the API around this principle.

Plus, it just makes sense because the SDL backend and GLFW backend are horribly inconsistent on when events actually happen, though this has improved recently it certainly is not 100% due to how they work. This would fix a number of inconsistencies reported by users between the two backends.

This makes more code common and reduces the responsibility of each individual backend, which will result in a more bug-free result and expectations will be more consistent: the input API is basically "driven" by the input context itself which doesn't know anything about the backend implementations, which to be honest is how we should've done this from the start.

Reply:
That does clear many things up for me; are users supposed to be able to simply copy the `device.State` struct to capture states? I still dislike the replacement of `KeyChar` though

Reply:
Yes (though a copy may not be mandatory as the property I'm hoping will be `ref readonly`, which will prevent the compiler from copying unless absolutely necessary).

Reply:
I've implemented the `ref readonly` now to prevent accidental copying.